src/common/checksumcalculator.cpp

```diff
@@ -0,0 +1,185 @@
+    } else if (checksumTypeName == checkSumSHA2C) {
+        _algorithmType = AlgorithmType::SHA256;
+    } else if (checksumTypeName == checkSumSHA3C) {
+        _algorithmType = AlgorithmType::SHA3_256;
+    } else if (checksumTypeName == checkSumAdlerC) {
+        _algorithmType = AlgorithmType::Adler32;
+    }
+
+    initChecksumAlgorithm();
+}
+
+ChecksumCalculator::~ChecksumCalculator()
+{
+    QMutexLocker locker(&_deviceMutex);
+    if (_device && _device->isOpen()) {
+        _device->close();
+    }
+}
+
+QByteArray ChecksumCalculator::calculate()
+{
+    QByteArray result;
+
+    if (!_isInitialized) {
+        return result;
+    }
+
+    Q_ASSERT(!_device->isOpen());
+    if (_device->isOpen()) {
+        qWarning() << "Device already open. Ignoring.";
+    }
+
+    if (!_device->isOpen() && !_device->open(QIODevice::ReadOnly)) {
+        if (auto file = qobject_cast<QFile *>(_device.data())) {
+            qWarning() << "Could not open file" << file->fileName() << "for reading to compute a checksum" << file->errorString();
+        } else {
+            qWarning() << "Could not open device" << _device.data() << "for reading to compute a checksum" << _device->errorString();
+        }
+        return result;
+    }
+
+    bool isAnyChunkAdded = false;
+
+    for (;;) {
+        QMutexLocker locker(&_deviceMutex);
+        if (!_device->isOpen() || _device->atEnd()) {
+            break;
+        }
+        const auto toRead = qMin(_device->bytesAvailable(), bufSize);
+        if (toRead <= 0) {
+            break;
+        }
+        QByteArray buf(toRead, Qt::Uninitialized);
+        const auto sizeRead = _device->read(buf.data(), toRead);
+        if (sizeRead <= 0) {
+            break;
+        }
+        if (!addChunk(buf, sizeRead)) {
+            break;
+        }
+        isAnyChunkAdded = true;
+    }
+
+    {
+        QMutexLocker locker(&_deviceMutex);
+        if (!_device->isOpen()) {
+            return result;
+        }
+    }
+
+    if (isAnyChunkAdded) {
+        if (_algorithmType == AlgorithmType::Adler32) {
+            result = QByteArray::number(_adlerHash, 16);
+        } else {
+            Q_ASSERT(_cryptographicHash);
+            if (_cryptographicHash) {
+                result = _cryptographicHash->result().toHex();
+            }
+        }
+    }
+
+    {
+        QMutexLocker locker(&_deviceMutex);
+        if (_device->isOpen()) {
+            _device->close();
+        }
+    }
+
+    return result;
+}
+
+void ChecksumCalculator::initChecksumAlgorithm()
+{
+    if (_algorithmType == AlgorithmType::Undefined) {
+        return;
+    }
+
+    {
+        QMutexLocker locker(&_deviceMutex);
+        if (_device->size() == 0) {
+            return;
+        }
+    }
+
+    if (_algorithmType == AlgorithmType::Adler32) {
+        _adlerHash = adler32(0L, Z_NULL, 0);
+    } else {
+        _cryptographicHash.reset(new QCryptographicHash(algorithmTypeToQCryptoHashAlgorithm(_algorithmType)));
```

Comment:
```suggestion
        _cryptographicHash.reset(QSharedPointer<QCryptographicHash>::create(algorithmTypeToQCryptoHashAlgorithm(_algorithmType));
```

Reply:
@mgallien this does not compile, I have `_cryptographicHash` as `QScopedPointer`

Reply:
OK